### ISSUE TITLE
Illegal start of statement

### ISSUE THREAD

**Issue by 4lex1v** [scalameta/paradise#160]
Original issue [#603](https://github.com/scalameta/scalameta/issues/603)
[Repo](https://github.com/4lex1v/MetaBlockIssue) with an example project
---
I'm trying to expand a function declaration into a private class and it's factory function:

```scala 
object test {
  @constr def Print(s: String): Call[Unit] 
}
```
where `@constr` should expand into a `private class` with a `def`, so the annotation result value looks this: `q"$genClass; $getFunc"`. It fails to compile:

```
[error] <macro>:2: illegal start of statement (no modifiers allowed here)
[error]   private final class $constr$PrintLine(val value: String) extends SystemCall[Unit] {}
[error]   ^
```

So it looks something like this after expansion:

```scala
object test {
  {
    private class generatedClass
    def generatedFunc
  }
}
```

maybe i'm simply missing some restriction. 

**Note**: 
1) The example project uses Typelevel Scala compiler, but Lightbend's doesn't work as well
2) Intellij's macro expansion feature works, but SBT fails to compile

**Comment by olafurpg** [comment]
Have you tried removing the private modifier? I think the class is already private since it's wrapped in a block.

**Comment by 4lex1v** [comment]
@olafurpg it works without the modifier, but i need to have it hidden within the `object`, hence i need to use private.

**Comment by olafurpg** [comment]
@4lex1v private classes are not allowed in Scala

```scala
scala> object a { { private class b } }
<console>:1: error: illegal start of statement (no modifiers allowed here)
object a { { private class b } }
             ^
```

could you get away with putting the class inside the `def` body?

**Comment by 4lex1v** [comment]
@olafurpg They are, but i don't need it inside the block, i want to generate a private class with the corresponding function, i.e:

```scala 
object outer {
  @constr def Something(): Trait[A]
}
```

into 

```scala
object outer {
  private class impl extends Trait[A]
  def Something() = ...
}
```

**Comment by olafurpg** [comment]
Would this work?

```scala
object outer {
  def Something() = {
    class impl extends Trait[A]
    ...
  }
}
```

**Comment by olafurpg** [comment]
I'm not sure if it's currently possible to expand as you describe, @DavidDudson may know better than me

**Comment by 4lex1v** [comment]
@olafurpg Unfortunately not. I wonder if described behaviour is intentional or it's a issue and may be eventually fixed?

**Comment by olafurpg** [comment]
I think the issue here is that `inline def apply` returns a `Tree` which cannot be two sequential statements unless it's wrapped in block. It might be possible to change it to return `Seq[Tree]`. I recommend you read over https://github.com/scalameta/paradise/issues/158 and pitch your thoughts

**Comment by 4lex1v** [comment]
@olafurpg i'm sorry but i don't understand, it perfectly works if i drop `private` modifier, but not with it?

**Comment by DavidDudson** [comment]
I'm wondering if this pattern match is failing for some reason.

https://github.com/scalameta/paradise/blob/master/plugin/src/main/scala/org/scalameta/paradise/typechecker/Expanders.scala#L202

I'll investigate tonight or tomorrow

**Comment by DavidDudson** [comment]
@olafurpg It looks like we definitely can switch to immutable.Seq[Tree], If I'm reading the expander code correctly

**Comment by xeno-by** [comment]
@DavidDudson How about treating blocks as sequences of trees by unwrapping them internally in expanders?

**Comment by DavidDudson** [comment]
@xeno-by That's what should be happening already. See that line of code I linked.

**Comment by xeno-by** [comment]
Yeah, that's how I was remembering it. I also was surprised that it didn't work.

**Comment by 4lex1v** [comment]
Does the issue with the `private` modifier related to the `Seq[Tree]` result type?

**Comment by DavidDudson** [comment]
Information on the issue:

This an example output of a macro which exibits this issue (Which is fine, it makes sense)

```scala
{
  private class Foo
  trait Bar
}
```

The problem, is that in order to get from that m.Tree, to the g.Tree

we of course call .syntax, and then reparse.

After the reparse we check the g.Tree for blocks and extract stats from it.

Its the reparse that breaks, I presume anything that it restricted during parse inside blocks (including private modifiers) will break.

**Comment by DavidDudson** [comment]
Fixed in #161

**Comment by DavidDudson** [comment]
Note private[X] still will not work
